src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -96,19 +111,19 @@ func (vmi *VMInterpreter_I) ApplyMessage(inTree st.StateTree, message msg.Unsign
 	)
 
 	if !sendRet.ExitCode().AllowsStateUpdate() {
-		// error -- revert all state changes -- ie drop updates. burn used gas.
+		// error -- revert all state changes -- ie drop updates.
 		outTree = inTree
-		outTree = _withTransferFundsAssert(
-			outTree,
-			message.From(),
-			addr.BurntFundsActorAddr,
-			gasToFIL(sendRet.GasUsed(), message.GasPrice()),
-		)
+
+		TODO() // TODO: As currently specced, this will burn the entire gas limit,
+		// not just used gas; is this desired? If not, should refund unused gas as below.
+
 	} else {
 		// success -- refund unused gas
 		outTree = sendRetStateTree
-		refundGas := message.GasLimit() - sendRet.GasUsed()
-		TODO() // TODO: assert refundGas is nonnegative
+		refundGas := message.GasLimit().Subtract(sendRet.GasUsed())
```

Comment:
This is somewhat tricky since the runtime has a separate `gasRemaining` counter, and should already be handling out-of-gas errors if they occur (i.e., an out-of-gas error _after_ the runtime call is an interpreter logic error). Will give this some thought to factor it in a way that's potentially easier to reason about.